### PR TITLE
[0.18.0] DruidInputSource can add new dimensions during re-ingestion (#9590)

### PR DESCRIPTION
**Pull request by suneet-s**
Backports the following commits to 0.18.0:
 - DruidInputSource can add new dimensions during re-ingestion (#9590)

### CODE REVIEW

**Comment by suneet-s** [comment]
clean backport

**Comment by maytasm** [comment]
Looks like you have COMPILATION ERROR from QUICKSTART_COMPATIBLE that was added in master

File: integration-tests/src/test/java/org/apache/druid/tests/indexer/ITTransformTest.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+import org.apache.druid.testing.guice.DruidTestModuleFactory;
+import org.apache.druid.tests.TestNGGroup;
+import org.testng.annotations.Guice;
+import org.testng.annotations.Test;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+@Test(groups = {TestNGGroup.BATCH_INDEX, TestNGGroup.QUICKSTART_COMPATIBLE})
```
Comment (suneet-s) [inline integration-tests/src/test/java/org/apache/druid/tests/indexer/ITTransformTest.java:30]:
```suggestion
@Test(groups = {TestNGGroup.BATCH_INDEX})
```

**Comment by suneet-s** [comment]
> Looks like you have COMPILATION ERROR from QUICKSTART_COMPATIBLE that was added in master
> Suggest removing the QUICKSTART_COMPATIBLE group

👍 thanks for pointing that out

**Pull request merged**